### PR TITLE
Add ARC MPU v3 & v6 support

### PR DESCRIPTION
**Pull request by YuguoWH**
Adding ARC MPU v3 and v6 support. 
The v3 uses same logic with v2 and the only difference is to support 32 bytes minimal region size. 
The v6 support 32 bytes, and it can have up to 32 regions. The regions are divided into 2 banks and switched by set/unset RB bit in MPU_EN aux register
This PR also introduces a new nsim hs board simulator which can validate/test with mpu v6.

### CODE REVIEW

**Comment by YuguoWH** [comment]
related issue https://github.com/zephyrproject-rtos/zephyr/issues/37309

File: include/arch/arc/arch.h [hunk 1]
```diff
@@ -67,6 +67,8 @@ extern "C" {
 #define Z_ARC_MPU_ALIGN	2048
 #elif CONFIG_ARC_MPU_VER == 4
 #define Z_ARC_MPU_ALIGN	32
+#elif CONFIG_ARC_MPU_VER == 6
+#define Z_ARC_MPU_ALIGN 32
```
Comment (abrodkin) [inline include/arch/arc/arch.h:71]: Maybe
```
#if (CONFIG_ARC_MPU_VER == 4) || (CONFIG_ARC_MPU_VER == 6)
```
?

File: include/arch/arc/v2/linker.ld [hunk 2]
```diff
@@ -32,6 +32,8 @@
 		#define MPU_MIN_SIZE 2048
 	#elif CONFIG_ARC_MPU_VER == 4
 		#define MPU_MIN_SIZE 32
+	#elif CONFIG_ARC_MPU_VER == 6
+		#define MPU_MIN_SIZE 32
```
Comment (abrodkin) [inline include/arch/arc/v2/linker.ld:36]: Ditto

File: arch/arc/core/mpu/arc_mpu_v6_internal.h [hunk 3]
```diff
@@ -0,0 +1,455 @@
+/*
+ * Copyright (c) 2019 Synopsys.
```
Comment (abrodkin) [inline arch/arc/core/mpu/arc_mpu_v6_internal.h:2]: 2021 ;)

File: arch/arc/core/mpu/arc_mpu_v2_internal.h [hunk 4]
```diff
@@ -17,7 +17,12 @@
 #define AUX_MPU_RDP_SIZE_MASK (0xE03)
 
 /* For MPU version 2, the minimum protection region size is 2048 bytes */
+#if CONFIG_ARC_MPU_VER == 2
 #define ARC_FEATURE_MPU_ALIGNMENT_BITS 11
+/* For MPU version 3, the minimum protection region size is 32 bytes */
+#else
```
Comment (abrodkin) [inline arch/arc/core/mpu/arc_mpu_v2_internal.h:23]: Maybe do explicit check for versions 3 & 6 to get rid of ambiguity?

File: arch/arc/core/mpu/arc_mpu_v6_internal.h [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_ARCH_ARC_CORE_MPU_ARC_MPU_V2_INTERNAL_H_
+#define ZEPHYR_ARCH_ARC_CORE_MPU_ARC_MPU_V2_INTERNAL_H_
```
Comment (abrodkin) [inline arch/arc/core/mpu/arc_mpu_v6_internal.h:7]: `ZEPHYR_ARCH_ARC_CORE_MPU_ARC_MPU_V6_INTERNAL_H_` (see **v6**)

File: arch/arc/core/mpu/arc_mpu_v6_internal.h [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+ * type (i.e. THREAD_STACK_REGION) and return the correct region index.
+ */
+static inline int get_region_index_by_type(uint32_t type)
+{
+	/*
+	 * The new MPU regions are allocated per type after the statically
+	 * configured regions. The type is one-indexed rather than
+	 * zero-indexed.
+	 *
+	 * For ARC MPU v2, the smaller index has higher priority, so the
```
Comment (abrodkin) [inline arch/arc/core/mpu/arc_mpu_v6_internal.h:78]: Why mention MPUv2 in code for MPUv6?

File: arch/arc/core/mpu/arc_mpu_v6_internal.h [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+
+/**
+ * @brief validate the given buffer is user accessible or not
+ */
+int arc_core_mpu_buffer_validate(void *addr, size_t size, int write)
+{
+	int r_index;
+
+	/*
+	 * For ARC MPU v2, smaller region number takes priority.
```
Comment (abrodkin) [inline arch/arc/core/mpu/arc_mpu_v6_internal.h:369]: Ditto

**Review by abrodkin — CHANGES_REQUESTED** [review]
Maybe move common functionality into version-agnostic file and then do specific things in v2 & v6 files to reduce duplication of code?

Also see some minor comments.

File: arch/arc/core/mpu/arc_mpu_v6_internal.h [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+	_bank_select(bank);
+	r_addr_start = z_arc_v2_aux_reg_read(_ARC_V2_MPU_RDB0 + index)
+		       & (~AUX_MPU_RDB_VALID_MASK);
+	r_size_lshift = z_arc_v2_aux_reg_read(_ARC_V2_MPU_RDP0 + index)
+			& AUX_MPU_RDP_SIZE_MASK;
+	r_size_lshift = (r_size_lshift & 0x3) | ((r_size_lshift >> 7) & 0x1C);
+	r_addr_end = r_addr_start  + (1 << (r_size_lshift + 1));
+
+	if (start >= r_addr_start && (start + size) <= r_addr_end) {
+		return 1;
```
Comment (evgeniy-paltsev) [inline arch/arc/core/mpu/arc_mpu_v6_internal.h:134]: return true;

**Review by evgeniy-paltsev — CHANGES_REQUESTED** [review]
Hi Yuguo,

please find some comments (mostly minor).

File: arch/arc/core/mpu/arc_mpu_v6_internal.h [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+/**
+ * This internal function checks if region is enabled or not
+ */
+static inline bool _is_enabled_region(uint32_t r_index)
+{
+	uint32_t bank = r_index / ARC_FEATURE_MPU_BANK_SIZE;
+	uint32_t index = (r_index % ARC_FEATURE_MPU_BANK_SIZE) * 2U;
+
+	_bank_select(bank);
+	return ((z_arc_v2_aux_reg_read(_ARC_V2_MPU_RDB0 + index)
```
Comment (abrodkin) [inline arch/arc/core/mpu/arc_mpu_v6_internal.h:109]: Maybe explicitly return some `bool`?

File: arch/arc/core/mpu/arc_mpu_v6_internal.h [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+#define ZEPHYR_ARCH_ARC_CORE_MPU_ARC_MPU_V6_INTERNAL_H_
+
+#define AUX_MPU_RDB_VALID_MASK (0x1)
+#define AUX_MPU_EN_ENABLE   (0x40000000)
+#define AUX_MPU_EN_DISABLE  (0xBFFFFFFF)
+#define AUX_MPU_EN_BANK_MASK (0x1)
+
+#define AUX_MPU_RDP_REGION_SIZE(bits)   (((bits - 1) & 0x3) | (((bits - 1) & 0x1C) << 7))
+
+#define AUX_MPU_RDP_ATTR_MASK (0x1F8)
```
Comment (abrodkin) [inline arch/arc/core/mpu/arc_mpu_v6_internal.h:16]: Frankly I don't like these cryptic numbers. I guess these are some groups of bits shifted left?
Look at `BIT_MASK()` in https://github.com/zephyrproject-rtos/zephyr/blob/main/include/sys/util_macro.h#L68. Then we'll get something like `BIT_MASK(xxx) << yyy` which is much more readable.

File: arch/arc/core/mpu/arc_mpu_v6_internal.h [hunk 11]
```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2021 Synopsys.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_ARCH_ARC_CORE_MPU_ARC_MPU_V6_INTERNAL_H_
+#define ZEPHYR_ARCH_ARC_CORE_MPU_ARC_MPU_V6_INTERNAL_H_
+
+#define AUX_MPU_RDB_VALID_MASK (0x1)
+#define AUX_MPU_EN_ENABLE   (0x40000000)
```
Comment (abrodkin) [inline arch/arc/core/mpu/arc_mpu_v6_internal.h:10]: Maybe `BIT(x)` and `~BIT(x)` correspondigly?

File: arch/arc/core/mpu/arc_mpu_v6_internal.h [hunk 12]
```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2021 Synopsys.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_ARCH_ARC_CORE_MPU_ARC_MPU_V6_INTERNAL_H_
+#define ZEPHYR_ARCH_ARC_CORE_MPU_ARC_MPU_V6_INTERNAL_H_
+
+#define AUX_MPU_RDB_VALID_MASK (0x1)
```
Comment (abrodkin) [inline arch/arc/core/mpu/arc_mpu_v6_internal.h:9]: `BIT(0)`?

**Review by abrodkin — APPROVED** [review]
With a couple of style nitpicks LGTM.

File: arch/arc/core/mpu/arc_mpu_v6_internal.h (hunk 9 above)
Comment (evgeniy-paltsev) [reply arch/arc/core/mpu/arc_mpu_v6_internal.h:109]: We are returning bool here (two values comparation result) but it is hidden as we split it for two lines...

**Review by ruuddw — APPROVED** [review]

File: arch/arc/core/mpu/arc_mpu_common_internal.h [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+{
+	int r_index;
+
+	/*
+	 * For ARC MPU, smaller region number takes priority.
+	 * we can stop the iteration immediately once we find the
+	 * matched region that grants permission or denies access.
+	 *
+	 */
+	for (r_index = 0; r_index < get_num_regions(); r_index++) {
```
Comment (evgeniy-paltsev) [inline arch/arc/core/mpu/arc_mpu_common_internal.h:221]: As we use r_index only in loop probably it's better to define it in loop itself:
```
for (int r_index = 0; r_index < get_num_regions(); r_index++) {
```

File: arch/arc/core/mpu/arc_mpu_common_internal.h [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+	int r_index;
+	/*
+	 * the MPU regions are filled in the reverse order.
+	 * According to ARCv2 ISA, the MPU region with smaller
+	 * index has higher priority. The static background MPU
+	 * regions in mpu_config will be in the bottom. Then
+	 * the special type regions will be above.
+	 *
+	 */
+	r_index = num_regions - mpu_config.num_regions;
```
Comment (evgeniy-paltsev) [inline arch/arc/core/mpu/arc_mpu_common_internal.h:272]: I don't think we need to split definition and declaration here. Let's combine it:
```
/* comment */
int r_index = num_regions - mpu_config.num_regions;
```

File: arch/arc/core/mpu/arc_mpu_v6_internal.h [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+	 *
+	 * For ARC MPU v6, the smaller index has higher priority, so the
+	 * index is allocated in reverse order. Static regions start from
+	 * the biggest index, then thread related regions.
+	 *
+	 */
+	switch (type) {
+	case THREAD_STACK_USER_REGION:
+		return get_num_regions() - mpu_config.num_regions
+		       - THREAD_STACK_REGION;
```
Comment (evgeniy-paltsev) [inline arch/arc/core/mpu/arc_mpu_v6_internal.h:85]: We don't exceed the Zephyr line limit here (100 symbols per line), please keep it in one line.

Please check rest of the code - we have a lot of places where we can put the whole expression in one line. It improves code readability a lot.

File: arch/arc/core/mpu/arc_mpu_v6_internal.h [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+{
+	uint32_t r_addr_start;
+	uint32_t r_addr_end;
+	uint32_t r_size_lshift;
+	uint32_t bank = r_index / ARC_FEATURE_MPU_BANK_SIZE;
+	uint32_t index = (r_index % ARC_FEATURE_MPU_BANK_SIZE) * 2U;
+
+	_bank_select(bank);
+	r_addr_start = z_arc_v2_aux_reg_read(_ARC_V2_MPU_RDB0 + index)
+		       & (~AUX_MPU_RDB_VALID_MASK);
```
Comment (evgeniy-paltsev) [inline arch/arc/core/mpu/arc_mpu_v6_internal.h:126]: Ditto.

File: arch/arc/core/mpu/arc_mpu_v6_internal.h [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+	r_size_lshift = z_arc_v2_aux_reg_read(_ARC_V2_MPU_RDP0 + index)
+			& AUX_MPU_RDP_SIZE_MASK;
```
Comment (evgeniy-paltsev) [inline arch/arc/core/mpu/arc_mpu_v6_internal.h:128]: Ditto.

File: arch/arc/core/mpu/arc_mpu_v6_internal.h [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+	r_size_lshift = (r_size_lshift & 0x3) | ((r_size_lshift >> 7) & 0x1C);
```
Comment (evgeniy-paltsev) [inline arch/arc/core/mpu/arc_mpu_v6_internal.h:129]: Should we move those magic numbers to some defines / constants?

File: arch/arc/core/mpu/arc_mpu_common_internal.h [hunk 19]
```diff
@@ -0,0 +1,300 @@
...
+
+	/* clear all the regions first */
+	for (i = 0U; i < r_index; i++) {
```
Comment (evgeniy-paltsev) [inline arch/arc/core/mpu/arc_mpu_common_internal.h:275]: As we use `i` only in loop probably it's better to define it in loop itself:
```
for (uint32_t i = 0U; i < r_index; i++) {
```

File: arch/arc/core/mpu/arc_mpu_common_internal.h [hunk 20]
```diff
@@ -0,0 +1,300 @@
...
+		_region_init(i, 0, 0, 0);
+	}
+
+	/* configure the static regions */
+	for (i = 0U; i < mpu_config.num_regions; i++) {
```
Comment (evgeniy-paltsev) [inline arch/arc/core/mpu/arc_mpu_common_internal.h:280]: Ditto.

**Review by evgeniy-paltsev — CHANGES_REQUESTED** [review]

**Review by abrodkin — CHANGES_REQUESTED** [review]
In commit titles it's better to say "add support _of_" instead of "to".
Also some comments below.

**Review by abrodkin — APPROVED** [review]
With 2 nitpicks LGTM

**Review by evgeniy-paltsev — APPROVED** [review]

**Review by nashif — APPROVED** [review]

File: soc/arc/snps_nsim/Kconfig [hunk 21]
```diff
@@ -29,6 +29,11 @@ config SOC_NSIM_HS_SMP
 	bool "Multi-core Synopsys ARC HS in nSIM"
 	select CPU_HAS_FPU
 
+config SOC_NSIM_HS_MPUV6
+	bool "Synopsys ARC HS with MPU v6 in nSIM"
+	select CPU_HAS_MPU
+	select CPU_HAS_FPU
```
Comment (ruuddw) [inline soc/arc/snps_nsim/Kconfig:35]: Why is FPU enabled?

Comment (YuguoWH) [reply soc/arc/snps_nsim/Kconfig:35]: I keep it same with other NSIM_HS boards, so FPU is enabled

**Pull request merged**